lib/ts/MemArena.cc

```diff
@@ -0,0 +1,179 @@
+  size      = 0;
+  allocated = 0;
+  data      = nullptr;
+  next      = nullptr;
+}
+
+/**
+    Marks a new generation. Caller should set next_block_size before creating a new generation.
+    Previous generation is implicitly frozen.
+ */
+inline void
+MemArena::create_new_generation()
+{
+  if (!next_block_size) {
+    next_block_size = arena_size * 2;
+  }
+
+  MemArenaBlock *next_gen = new MemArenaBlock();
+  next_gen->alloc(next_block_size);
+
+  // update generation marker
+  generation->next = next_gen;
+  generation       = next_gen;
+  arena_size       = next_block_size;
+}
+
+MemArena::MemArena()
+{
+  MemArena(arena_size); // Default size
+}
+
+MemArena::MemArena(size_t n)
+{
+  allocations = new MemArenaBlock();
+  allocations->alloc(n);
+  arena_size = n;
+  generation = allocations; // first generation
+}
+
+MemArena::~MemArena()
+{
+  MemArenaBlock *tmp = allocations;
+  MemArenaBlock *next;
+  while (tmp) {
+    next = tmp->next;
+    tmp->free();
+    delete tmp;
+
+    tmp = next;
+  }
+
+  allocations = nullptr;
+  generation  = nullptr;
+}
+
+MemSpan
+MemArena::alloc(size_t n)
+{
+  std::lock_guard<std::mutex> lock(arena_mutex);
+
+  // frozen. must allocate in a new generation.
+  if (next_generation) {
+    create_new_generation();
+    next_generation = false;
+  }
+
+  // keep creating generations memory blocks until we have something large enough
+  while (generation->size - generation->allocated /* remaining size */ < n) {
+    next_block_size = arena_size * 2;
+    create_new_generation();
+  }
+
+  ink_assert(generation->size >= n);
+
+  uint64_t offset = generation->allocated;
+  generation->allocated += n;
+
+  // Allocate a span of memory within the block.
+  MemSpan ret((char *)generation->data + offset, n);
+  return ret;
+}
+
+MemArena &
+MemArena::freeze(size_t n)
+{
+  // Get the mutex here to stop other allocations in internal memory blocks.
+  std::lock_guard<std::mutex> lock(arena_mutex);
+
+  next_generation = true;
+  next_block_size = n ? n : arena_size * 2; // If no size specified, just double in size
+
+  return *this;
+}
+
+/**
+    Everything up the current generation is considered frozen
+     as it is impossible to allocate to those memory blocks.
+ */
+MemArena &
+MemArena::thaw()
+{
+  MemArenaBlock *tmp = allocations;
+  MemArenaBlock *next;
+  while (tmp) {
+    if (tmp == generation)
+      break;
+
+    // frozen block, deallocate
+    next = tmp->next;
+    tmp->free();
+    delete tmp;
+    tmp = next;
+  }
+
+  allocations = generation;
+  return *this;
+}
+
+bool
+MemArena::contains(void *ptr) const
+{
+  // Since the internal memory blocks can be allocated at different
+  // times, there's no guarantee that the underlying memory is continuous.
+  // Therefore, we must search through each of the generations individually.
+
+  MemArenaBlock *tmp = allocations;
+  while (tmp) {
+    if (ptr >= static_cast<char *>(tmp->data) && ptr < static_cast<char *>(tmp->data) + tmp->size) {
```

Comment:
The `Block` class should have a `data()` method that returns a `char*` to its allocated data.